### PR TITLE
SDPA integration for nvFuser

### PR DESCRIPTION
**Pull request by Priya2698**
This PR:
1. Adds SDPA operator for nvFuser -- currently it is behind the `nv_enable_sdpa=True` compile flag, and only supports flash attention.
2. Moves `register_operator` to `Executor` class to allow use for `nvFuserExecutor`.
3. Moves some common utility functions from `sdpaex.py` to `utils.py`

The motivation behind this operator is to support multi-GPU development in nvFuser. At present, we do not have plans to generate code for this operator. It will be executed through ATen fallback path in nvFuser.

See https://github.com/NVIDIA/Fuser/issues/2278 for discussions.

CC: @jjsjann123 @wujingyue

### CODE REVIEW

File: thunder/__init__.py [hunk 1]
```diff
@@ -183,11 +183,11 @@ def __version__():
...
 
-if sdpa_executor:
-    add_default_executor(sdpa_executor)
+# if sdpa_executor:
+#     add_default_executor(sdpa_executor)
 
-if cudnn_executor:
-    add_default_executor(cudnn_executor)
+# if cudnn_executor:
+#     add_default_executor(cudnn_executor)
```
Comment (Priya2698) [inline thunder/__init__.py:190]: Investigating why SDPA is passed to cudnn by default in nvfuser tests as well where the executor is specified.

**Comment by wujingyue** [comment]
Cool! I'll try it with https://github.com/NVIDIA/Fuser/blob/13ed2bc60836c107d3edfddc5c1bfc4deb79ffb8/benchmarks/python/test_transformer.py#L5-L39 next week.

**Comment by wujingyue** [comment]
It seems to work for me! At least, I'm seeing one fusion for forward and one for backward, and they both contain sdpfa nodes. 

```
$ grep fusion_id /tmp/stdout
def nvfuser_fusion_id1(fd : FusionDefinition) -> None :
    nvfuser_fusion_id1(fd)
def nvfuser_fusion_id2(fd : FusionDefinition) -> None :
    nvfuser_fusion_id2(fd)
def nvfuser_fusion_id2(fd : FusionDefinition) -> None :
    nvfuser_fusion_id2(fd)
def nvfuser_fusion_id2(fd : FusionDefinition) -> None :
    nvfuser_fusion_id2(fd)
def nvfuser_fusion_id2(fd : FusionDefinition) -> None :
    nvfuser_fusion_id2(fd)
def nvfuser_fusion_id2(fd : FusionDefinition) -> None :
    nvfuser_fusion_id2(fd)
def nvfuser_fusion_id2(fd : FusionDefinition) -> None :
    nvfuser_fusion_id2(fd)
def nvfuser_fusion_id2(fd : FusionDefinition) -> None :
    nvfuser_fusion_id2(fd)
def nvfuser_fusion_id2(fd : FusionDefinition) -> None :
    nvfuser_fusion_id2(fd)
def nvfuser_fusion_id2(fd : FusionDefinition) -> None :
    nvfuser_fusion_id2(fd)
def nvfuser_fusion_id2(fd : FusionDefinition) -> None :
    nvfuser_fusion_id2(fd)
def nvfuser_fusion_id2(fd : FusionDefinition) -> None :
    nvfuser_fusion_id2(fd)
def nvfuser_fusion_id2(fd : FusionDefinition) -> None :
    nvfuser_fusion_id2(fd)
def nvfuser_fusion_id2(fd : FusionDefinition) -> None :
    nvfuser_fusion_id2(fd)
def nvfuser_fusion_id2(fd : FusionDefinition) -> None :
    nvfuser_fusion_id2(fd)
def nvfuser_fusion_id2(fd : FusionDefinition) -> None :
    nvfuser_fusion_id2(fd)
def nvfuser_fusion_id2(fd : FusionDefinition) -> None :
    nvfuser_fusion_id2(fd)
def nvfuser_fusion_id2(fd : FusionDefinition) -> None :
    nvfuser_fusion_id2(fd)
def nvfuser_fusion_id2(fd : FusionDefinition) -> None :
    nvfuser_fusion_id2(fd)
def nvfuser_fusion_id2(fd : FusionDefinition) -> None :
    nvfuser_fusion_id2(fd)
def nvfuser_fusion_id2(fd : FusionDefinition) -> None :
    nvfuser_fusion_id2(fd)

$ grep sdpfa /tmp/stdout
    T87, T88, T89, T90 = fd.ops.sdpfa_fwd(T77, T70, T84, S85, S86, None)
    T336, T337, T338 = fd.ops.sdpfa_bwd(T333, T75, T68, T82, T7, T22, S334, S335, T23, T24, None)
    T336, T337, T338 = fd.ops.sdpfa_bwd(T333, T75, T68, T82, T7, T22, S334, S335, T23, T24, None)
    T336, T337, T338 = fd.ops.sdpfa_bwd(T333, T75, T68, T82, T7, T22, S334, S335, T23, T24, None)
    T336, T337, T338 = fd.ops.sdpfa_bwd(T333, T75, T68, T82, T7, T22, S334, S335, T23, T24, None)
    T336, T337, T338 = fd.ops.sdpfa_bwd(T333, T75, T68, T82, T7, T22, S334, S335, T23, T24, None)
    T336, T337, T338 = fd.ops.sdpfa_bwd(T333, T75, T68, T82, T7, T22, S334, S335, T23, T24, None)
    T336, T337, T338 = fd.ops.sdpfa_bwd(T333, T75, T68, T82, T7, T22, S334, S335, T23, T24, None)
    T336, T337, T338 = fd.ops.sdpfa_bwd(T333, T75, T68, T82, T7, T22, S334, S335, T23, T24, None)
    T336, T337, T338 = fd.ops.sdpfa_bwd(T333, T75, T68, T82, T7, T22, S334, S335, T23, T24, None)
    T336, T337, T338 = fd.ops.sdpfa_bwd(T333, T75, T68, T82, T7, T22, S334, S335, T23, T24, None)
    T336, T337, T338 = fd.ops.sdpfa_bwd(T333, T75, T68, T82, T7, T22, S334, S335, T23, T24, None)
    T336, T337, T338 = fd.ops.sdpfa_bwd(T333, T75, T68, T82, T7, T22, S334, S335, T23, T24, None)
    T336, T337, T338 = fd.ops.sdpfa_bwd(T333, T75, T68, T82, T7, T22, S334, S335, T23, T24, None)
    T336, T337, T338 = fd.ops.sdpfa_bwd(T333, T75, T68, T82, T7, T22, S334, S335, T23, T24, None)
    T336, T337, T338 = fd.ops.sdpfa_bwd(T333, T75, T68, T82, T7, T22, S334, S335, T23, T24, None)
    T336, T337, T338 = fd.ops.sdpfa_bwd(T333, T75, T68, T82, T7, T22, S334, S335, T23, T24, None)
    T336, T337, T338 = fd.ops.sdpfa_bwd(T333, T75, T68, T82, T7, T22, S334, S335, T23, T24, None)
    T336, T337, T338 = fd.ops.sdpfa_bwd(T333, T75, T68, T82, T7, T22, S334, S335, T23, T24, None)
    T336, T337, T338 = fd.ops.sdpfa_bwd(T333, T75, T68, T82, T7, T22, S334, S335, T23, T24, None)
    T336, T337, T338 = fd.ops.sdpfa_bwd(T333, T75, T68, T82, T7, T22, S334, S335, T23, T24, None)
    # (t43, t44, t45, t46) = nv_sdpfa_fwd(t34, t26, t42, 0.1, True, scale=None)
    # (t563, t564, t565) = nv_sdpfa_bwd(t562, t34, t26, t42, t43, t44, 0.1, True, t45, t46, scale=None)
```

File: thunder/__init__.py (hunk 1 above)
Comment (crcrpar) [reply thunder/__init__.py:190]: I think this should be done in a different PR.

File: thunder/executors/nvfuserex_impl.py [hunk 2]
```diff
@@ -2208,3 +2223,227 @@ def matmul(
...
+    query: TensorLike,
+    key: TensorLike,
+    value: TensorLike,
+    dropout_p: float = 0.0,
+    is_causal: bool = False,
+    *,
+    scale: None | float = None,
+) -> tuple[TensorProxy, TensorProxy, int, int]:
+    # Reference metadata:
+    # https://github.com/pytorch/pytorch/blob/main/torch/_meta_registrations.py
```
Comment (crcrpar) [inline thunder/executors/nvfuserex_impl.py:2246]: could you use a permalink instead? also, ideally could you have line info as well for a more solid reference in the future?

File: thunder/executors/nvfuserex_impl.py [hunk 3]
```diff
@@ -2208,3 +2223,227 @@ def matmul(
...
+    philox_offset: TensorLike,
+    *,
+    scale: None | float = None,
+) -> tuple[TensorProxy, TensorProxy, TensorProxy]:
+
+    batch_size, num_heads, query_seq_len, E = query.shape
+    key_seq_len = key.shape[2]
+
+    # Reference metadata:
+    # https://github.com/pytorch/pytorch/blob/main/torch/_meta_registrations.py#L4907-L4956
```
Comment (crcrpar) [inline thunder/executors/nvfuserex_impl.py:2316]: Could you make this a permalink?

**Review by crcrpar — COMMENTED** [review]
This looks nice to me, I'm wondering if you happen to have run some benchmark to compare the performance with and without nvfuser SDPA?

File: thunder/__init__.py (hunk 1 above)
Comment (Priya2698) [reply thunder/__init__.py:190]: These lines will be restored, this was for verifying nvFuser API.
Since cudnn, sdpa executor are always used as executors even when the executor list is specified, SDPA is not given to nvFuser. I will fix that before putting up this PR for review and merge.

**Comment by Priya2698** [comment]
> This looks nice to me, I'm wondering if you happen to have run some benchmark to compare the performance with and without nvfuser SDPA?

@wujingyue has a benchmark in this PR: https://github.com/NVIDIA/Fuser/pull/2787.
I will run some benchmarks later this week.

**Comment by wujingyue** [comment]
> > This looks nice to me, I'm wondering if you happen to have run some benchmark to compare the performance with and without nvfuser SDPA?
> 
> @wujingyue has a benchmark in this PR: [NVIDIA/Fuser#2787](https://github.com/NVIDIA/Fuser/pull/2787). I will run some benchmarks later this week.

I'm not sure if we are talking about the same thing. NVIDIA/Fuser#2787 compares SDPA vs the decomposed version of that (transpose, matmul, softmax, matmul etc). @crcrpar, is that what you were interested in? Or do you want to compare with and without Thunder's SDPA executor?

**Comment by crcrpar** [comment]
Excuse my unclear question, I'm interested in the performance benefit of using nvfuser's SDPA in thunder's benchmarks, e.g. https://github.com/Lightning-AI/lightning-thunder/blob/e88966f2f35433e167f40d26e1330684b8ee3fc1/thunder/benchmarks/benchmark_litgpt.py.

E.g., how would it affect the performance of llama-2-7b-hf / llama-3-8B with FSDP zero2?
An example command would be
```
$ torchrun --nproc-per-node=8 --local-ranks-filter 0 --role rank --tee 3 thunder/benchmarks/benchmark_litgpt.py --compile=thunder_inductor_cat --distributed_mode=fsdp --model_name=Llama-2-7b-hf
```

**Comment by Priya2698** [comment]
> Excuse my unclear question, I'm interested in the performance benefit of using nvfuser's SDPA in thunder's benchmarks, e.g. https://github.com/Lightning-AI/lightning-thunder/blob/e88966f2f35433e167f40d26e1330684b8ee3fc1/thunder/benchmarks/benchmark_litgpt.py.
> 
> E.g., how would it affect the performance of llama-2-7b-hf / llama-3-8B with FSDP zero2? An example command would be
> 
> ```
> $ torchrun --nproc-per-node=8 --local-ranks-filter 0 --role rank --tee 3 thunder/benchmarks/benchmark_litgpt.py --compile=thunder_inductor_cat --distributed_mode=fsdp --model_name=Llama-2-7b-hf
> ```

I am yet to run these end-to-end benchmarks.

File: thunder/__init__.py (hunk 1 above)
Comment (jjsjann123) [reply thunder/__init__.py:190]: Not trying to rush it, but is this PR ready for review yet?

File: thunder/executors/nvfuserex_impl.py (hunk 2 above)
Comment (Priya2698) [reply thunder/executors/nvfuserex_impl.py]: I borrowed this link from `sdpaex.py ` but this file does not have the metadata for `forward`. I modified this to be a descriptive comment.

**Review by jjsjann123 — APPROVED** [review]
LGTM. :partying_face:

File: thunder/tests/test_nvfuser.py [hunk 4]
```diff
@@ -991,3 +991,78 @@ def make_integer_tensor():
...
         jout = f(x, y)
         assert rout.equal(jout)
+
+
+@instantiate(
+    dtypes=(thunder.float16, thunder.bfloat16),
+    devicetypes=(devices.DeviceType.CUDA,),
+    executors=(nvFuserExecutor,),
+    decorators=(
+        pytest.mark.skipif(
```
Comment (rdspring1) [inline thunder/tests/test_nvfuser.py:1001]: I wonder if the `ops` decorator is more appropriate here, so we can reuse the existing `sdpa` opinfo.

Example with cudnn executor. https://github.com/Lightning-AI/lightning-thunder/blob/main/thunder/tests/framework.py#L362

File: thunder/executors/nvfuserex_impl.py [hunk 5]
```diff
@@ -2208,3 +2223,230 @@ def matmul(
...
+        return False
+
+    enable_sdpa: None | bool = get_compile_option("nv_enable_sdpa", "Enable nvFuser flash attention SDPA.")
+
+    if not enable_sdpa:
+        return False
+
+    if not are_supported_tensors(query, key, value):
+        return False
+
```
[the remark this answers is not in the document]
Comment (Priya2698) [reply thunder/executors/nvfuserex_impl.py:2389]: Added a check.

File: thunder/executors/nvfuserex_impl.py [hunk 6]
```diff
@@ -2208,3 +2223,230 @@ def matmul(
...
+    batch_size, num_heads, query_seq_len, E = query.shape
+    key_seq_len = key.shape[2]
+
+    return (
+        output := TensorProxy(like=query, shape=(batch_size, num_heads, query_seq_len, E)),
+        log_sumexp := TensorProxy(
+            shape=(batch_size, num_heads, query_seq_len), dtype=dtypes.float32, device=query.device, requires_grad=False
+        ),
+        philox_seed := TensorProxy(shape=(), dtype=dtypes.int64, device=query.device, requires_grad=False),
+        philox_offset := TensorProxy(shape=(), dtype=dtypes.int64, device=query.device, requires_grad=False),
```
[the remark this answers is not in the document]
Comment (Priya2698) [reply thunder/executors/nvfuserex_impl.py]: Good catch, fixed.

File: thunder/tests/test_nvfuser.py (hunk 4 above)
Comment (Priya2698) [reply thunder/tests/test_nvfuser.py:1001]: I considered that. I was concerned about there being no case where Flash Attention is the chosen backend (since they had different E, Ev) and the test silently not checking anything.
So created the test in the current format.

**Review by t-vi — APPROVED** [review]
Stamping Thank you @Priya2698  @jjsjann123 @wujingyue @crcrpar @rdspring1 @mruberry

File: thunder/tests/test_nvfuser.py (hunk 4 above)
Comment (IvanYashchuk) [inline thunder/tests/test_nvfuser.py:1001]: To streamline our internal QA process, every test that uses Flash Attention should be skipped on older GPUs using the compute capability check (https://github.com/Lightning-AI/lightning-thunder/pull/1327).

Comment (wujingyue) [reply thunder/tests/test_nvfuser.py:1001]: Thanks, @IvanYashchuk! Is there a way to avoid having to remember this? E.g. let CI run tests on older GPUs because apparently people still care about them.

Comment (IvanYashchuk) [reply thunder/tests/test_nvfuser.py:1001]: Currently, Thunder's CI uses Lightning's resources. We could add runners with NVIDIA's resources including older GPUs similarly to how nvFuser's CI is set up. I would postpone doing that until we hit the problem again.

Alternatively, the test (and any test that uses Flash Attention) can be rewritten with a parametrization that turns on and off Flash Attention in PyTorch and it should pass when the Flash Attention is turned off ([`torch.backends.cuda.enable_flash_sdp`](https://pytorch.org/docs/stable/backends.html#torch.backends.cuda.enable_flash_sdp)). `fused_sdp_choice` which nvFuser's checker uses would pick up the value of this flag
https://github.com/Lightning-AI/lightning-thunder/blob/947e9ef2eca5c1a1b3c380d38e16c0afd7533b06/thunder/executors/utils.py#L216
https://github.com/Lightning-AI/lightning-thunder/blob/947e9ef2eca5c1a1b3c380d38e16c0afd7533b06/thunder/executors/nvfuserex_impl.py#L2456

Comment (Priya2698) [reply thunder/tests/test_nvfuser.py:1001]: Thanks @IvanYashchuk  for the fix.
